### ISSUE TITLE
UserProfile Endpoint

### ISSUE THREAD

**Issue by vanwars** [LocalGround/localground#74]
We need to make an endpoint that allows a user to:
* Change display name (first / last name)
* Set their default location
* Indicate whether they want email notifications

This will update the UserPrefs model.

**Comment by awesomescot** [comment]
It looks like there is already a function to set default location in site/api/views/user_profile_views.py: line 40: def update_user_location(request):  Or am I wrong?  So I guess I could look to that for an example?  Also I'm unable to find UserPrefs model, I did find UserProfile model though, I'm guessing that what I'll be modifying?  Also, should this be a single endpoint, that takes json and changes 1 to 3 of the parameters?  Or should it be 3 endpoints?

**Comment by awesomescot** [comment]
After looking at this with Tap, it looks like most of those fields are already in there.  But we could be wrong.  I guess I just need a little more detail as to exactly what is required from this ticket.  My understanding is that we want these fields to be updatable.  So this form should have those fields in it https://dev.localground.org/api/0/user-profile/1/.  Is that correct?  Is that all this ticket is referring to?

**Comment by vanwars** [comment]
Hi Scot, I did not realize that there was already a user-profile endpoint. Yes, **UserProfile** is the model, not UserPrefs. Sorry this issue is confusing.

Read-only fields:
* id
* foreign key to User model

Read / Write fields:
* display name
* email_announcements (boolean yes/no)
* default_view_authority (dropdown / lookup)
* default_location (geoJSON field...see Photo serializer for more detail on how geoJSON fields work)

**Comment by awesomescot** [comment]
ok, so quick question.  It seems like the display name would come from the foreign key to User, something like User.name.  Or am I misinterpreting something?  Also I don't see a display name in the model.  Should I add it?

**Comment by tap2k** [comment]
I think the right approach is to figure out how to access/modify User.name from the User Profile endpoint.

**Comment by vanwars** [comment]
Yes, exactly. I think you can create a UserProfile model serializer, add a
custom field for first name and last name, and then override the
serializer's "create" and "update" methods to also modify the User object.
You can look at the barcoded_serializer.py as an example of how overrides
work (as you will probably recall). So after UserProfile save, you can do a
second commit off of the instance object something like:
```
self.instance.user.first_name = validated_data.get('first_name')
self.instance.user.last_name = validated_data.get('last_name')
self.instance.user.save()
```

**Comment by awesomescot** [comment]
Ok, I think I got the saving figured out, but following your instructions.  Now for the serializing.  I thought this would work.  first_name = serializers.CharField(source=user.first_name), but of course user is a HyperlinkedRelatedField object, not the model object.  So how do I access the model that is currently being serialized?

**Comment by awesomescot** [comment]
Ok, so I've got most things kind of working.  The first_name and last_name coming from the User model works, but I can't figure out the correct queryset.  And I'm not entirely sure what to do with the default_location, maybe it's correct right now.  Anyway, if someone could check out the userprofile branch and take a look at my code and give me some feed back, that would be nice.
